Add IQS004 rule (#2367)
home-assistant/developers.home-assistant: docs/core/integration-quality-scale/rules/iqs004.md

---
title: "IQS004: Full test coverage for the config flow"
---

IQS004 is as follows:

Full test coverage for the config flow

## Reasoning

The config flow is the first interaction a user has with your integration.
It is important to ensure that the config flow is working as expected and that the user can set up the integration without any issues or running into any (config flow related) errors.

This means that we want to have **100%** test coverage for the config flow.

> During those tests, we also require tests where we verify that the flow is able to recover from an error to confirm that the user is able to finish the flow even if something goes wrong.

Reviewer: ```suggestion
In those tests, we require verification that the flow is able to recover from an error to confirm that the user is able to finish the flow even if something goes wrong.
```

Since we want the user to have a smooth experience using other integration flows, this rule also applies to the reconfigure, reauthentication, and options flows.

The extra added benefit of having tests for an integration is that it introduces the developer to testing, making it easier to write tests for other parts of the integration.

:::warning

> Even though the code used to check the uniqueness of a config entry is most likely touched by the happy flow tests, we still require tests for the uniqueness check to reach **100%** coverage.

Reviewer: ```suggestion
Even though the code used to check the uniqueness of a config entry is most likely touched by the happy flow tests, make sure to also test that the flow doesn't allow adding more than one unique configuration entry to reach complete coverage.
```
